sdk/python/kfp/components/structures/components.json_schema.json

```diff
@@ -0,0 +1,402 @@
+{
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "$id": "http://kubeflow.org/pipelines/components.json_schema.json",
+
+  "$ref": "#/definitions/PipelineRunSpec",
+
+  "definitions": {
+    "PrimitiveTypes": {
+      "oneOf": [
+        {"type": "string"},
+        {"type": "integer"},
+        {"type": "number"},
+        {"type": "boolean"}
+      ]
+    },
+
+    "TypeType": {
+      "oneOf": [
+        {"type": "string"},
+        {"type": "array", "items": {"$ref": "#/definitions/TypeType"}},
+        {"type": "object", "additionalProperties": {"$ref": "#/definitions/TypeType"}}
+      ]
+    },
+
+    "InputSpec": {
+      "description": "Describes the component input specification",
+      "type": "object",
+      "properties": {
+        "name": {"type": "string"},
+        "type": {"$ref": "#/definitions/TypeType"},
+        "description": {"type": "string"},
+        "default": {"$ref": "#/definitions/PrimitiveTypes"},
+        "optional": {"type": "boolean", "default": false}
+      },
+      "additionalProperties": false
+    },
+
+    "OutputSpec": {
+      "description": "Describes the component output specification",
+      "type": "object",
+      "required": ["name"],
```

Comment:
I would propose that we don't reduce functionality compared to Argo on how parameters can be used to specify container specs.

Reply:
I'm quite open to this suggestion. Especially to implementing this support for the existing properties of `ContainerSpec` (BTW, did you add comment for a wrong line?).

However having such support for the whole deeply nested k8s Pod spec and Container spec is another story. I think I'm striking a good balance for the current development stage of the project without closing doors for future flexibility. I'd greatly appreciate real-world CUJs and examples of components where the flexibility we're giving the users is insufficient.

Generally I go from the user examples. At this moment these examples include out 1st party components, some future components we're building, some requests of our clients and also the component set I was familiar with at MS. Having more examples will help us make better decisions.